# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/HTMLEditorNestedClasses.h
# Commit: 5bacd687fa98
# Full Hash: 5bacd687fa98b0e9a5dea9944f360fdc5a6770e6
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-01-17 16:13:02
# Regressor Bug: 1807829
# File Overlap Count: 2
# Description:
#   Bug 1807829 - Make `HTMLEditor::CreateStyleForInsertText` use same logic as `HTMLEditor::SetInlinePropertiesAsSubAction` to apply new style r=m_kato
#   
#   The problem in Yahoo mail is, when you apply new font size to:
#   ```
#   <div><font size="7">{}<br></font></div>
# ==============================================================================

diff -r a7465d2d89ba -r 5bacd687fa98 editor/libeditor/HTMLEditorNestedClasses.h
--- a/editor/libeditor/HTMLEditorNestedClasses.h	Mon Jan 16 23:14:12 2023 +0000
+++ b/editor/libeditor/HTMLEditorNestedClasses.h	Tue Jan 17 05:26:14 2023 +0000
@@ -76,7 +76,8 @@
    * See comments in the definition what this does.
    */
   Result<EditorRawDOMRange, nsresult> ExtendOrShrinkRangeToApplyTheStyle(
-      const HTMLEditor& aHTMLEditor, const EditorDOMRange& aRange) const;
+      const HTMLEditor& aHTMLEditor, const EditorDOMRange& aRange,
+      const Element& aEditingHost) const;
 
   /**
    * Returns next/previous sibling of aContent or an ancestor of it if it's
@@ -87,6 +88,32 @@
   [[nodiscard]] static nsIContent* GetPreviousEditableInlineContent(
       const nsIContent& aContent, const nsINode* aLimiter = nullptr);
 
+  /**
+   * GetEmptyTextNodeToApplyNewStyle creates new empty text node to insert
+   * a new element which will contain newly inserted text or returns existing
+   * empty text node if aCandidatePointToInsert is around it.
+   *
+   * NOTE: Unfortunately, editor does not want to insert text into empty inline
+   * element in some places (e.g., automatically adjusting caret position to
+   * nearest text node).  Therefore, we need to create new empty text node to
+   * prepare new styles for inserting text.  This method is designed for the
+   * preparation.
+   *
+   * @param aHTMLEditor                 The editor.
+   * @param aCandidatePointToInsert     The point where the caller wants to
+   *                                    insert new text.
+   * @param aEditingHost                The editing host.
+   * @return            If this creates new empty text node returns it.
+   *                    If this couldn't create new empty text node due to
+   *                    the point or aEditingHost cannot have text node,
+   *                    returns nullptr.
+   *                    Otherwise, returns error.
+   */
+  [[nodiscard]] MOZ_CAN_RUN_SCRIPT static Result<RefPtr<Text>, nsresult>
+  GetEmptyTextNodeToApplyNewStyle(HTMLEditor& aHTMLEditor,
+                                  const EditorDOMPoint& aCandidatePointToInsert,
+                                  const Element& aEditingHost);
+
  private:
   [[nodiscard]] MOZ_CAN_RUN_SCRIPT Result<CaretPoint, nsresult> ApplyStyle(
       HTMLEditor& aHTMLEditor, nsIContent& aContent);